util/otbn_build.py

```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import logging as log
+import os
+import shlex
+import subprocess
+import sys
+from pathlib import Path
+from typing import List
+
+log.basicConfig(level=log.INFO, format="%(message)s")
+
+repo_top = (Path(os.path.dirname(__file__)) / '../').resolve()
+
+
+def run_cmd(cmd, **kwargs):
+    cmd_str = ' '.join([shlex.quote(str(a)) for a in cmd])
+    log.info(cmd_str)
+    subprocess.run(cmd, **kwargs)
+
+
+def call_otbn_as(src_file: Path, out_file: Path):
+    otbn_as_cmd = os.environ.get('OTBN_AS',
+                                 str(repo_top / 'hw/ip/otbn/util/otbn-as'))
+    run_cmd([otbn_as_cmd, '-o', out_file, src_file], check=True)
+
+
+def call_otbn_ld(src_files: List[Path], out_file: Path):
+    otbn_ld_cmd = os.environ.get('OTBN_LD',
+                                 str(repo_top / 'hw/ip/otbn/util/otbn-ld'))
+    run_cmd([otbn_ld_cmd, '-o', out_file] + src_files, check=True)
+
+
+def call_rv32_objcopy(args: List[str]):
+    rv32_tool_objcopy = os.environ.get('RV32_TOOL_OBJCOPY',
+                                       'riscv32-unknown-elf-objcopy')
+    run_cmd([rv32_tool_objcopy] + args, check=True)
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--out-dir',
+        '-o',
+        required=False,
+        default=".",
+        help="Output directory (default: %(default)s)")
+    parser.add_argument(
+        '--app-name',
+        '-n',
+        required=False,
+        help="Name of the application, used as basename for the output. "
+             "Default: basename of the first source file.")
+    parser.add_argument('src_files', nargs='+', type=str, metavar='SRC_FILE')
+    args = parser.parse_args()
+
+    out_dir = Path(args.out_dir)
+    out_dir.mkdir(exist_ok=True)
+
+    src_files = [Path(f) for f in args.src_files]
+    for src_file in src_files:
+        if not src_file.exists():
+            print("Source file %s not found." % src_file)
+            return 1
+    obj_files = [out_dir / f.with_suffix('.o').name for f in src_files]
+
+    if args.app_name is not None:
+        app_name = args.app_name
+    else:
+        app_name = str(src_files[0].stem)
+
+    try:
+        for i in range(len(src_files)):
+            call_otbn_as(src_files[i], obj_files[i])
+
+        out_elf = out_dir / (app_name + '.elf')
+        call_otbn_ld(obj_files, out_elf)
+
+        out_embedded_obj = out_dir / (app_name + '.rv32embed.o')
+        args = [
+            '-O',
+            'elf32-littleriscv',
+            '--prefix-symbols',
+            '_otbn_app_' + app_name + '_',
+            out_elf,
+            out_embedded_obj,
+        ]
+
+        call_rv32_objcopy(args)
+    except subprocess.CalledProcessError as e:
+        # Show a nicer error message if any of the called programs fail.
+        cmd_str = ' '.join([shlex.quote(str(a)) for a in e.cmd])
```

Comment:
This command-pretty-logging code is repeated in another function. Factor it out?